### PR TITLE
Add retrying and error logging dispatchers

### PR DESCRIPTION
**Pull request by jbradford**
Also added a SkipOrDie dispatcher.

### CODE REVIEW

File: src/core/ErrorLoggingDIspatcher.cs [hunk 1]
```diff
@@ -0,0 +1,29 @@
+﻿using System;
+
+namespace CR.MessageDispatch.Core
+{
+    public class ErrorLoggingDispatcher<TMessage> : IDispatcher<TMessage>
+    {
+        private readonly IDispatcher<TMessage> _dispatcher;
+        private readonly Action<TMessage, Exception> _exceptionLogAction;
```
Comment (rcknight) [inline src/core/ErrorLoggingDIspatcher.cs]: Wasn't this going to be Exception, String?

File: src/core/OnExceptionSkipOrExitDispatcher.cs [hunk 2]
```diff
@@ -0,0 +1,37 @@
...
+        private readonly IDispatcher<TMessage> _dispatcher;
+        private readonly Func<TMessage, Exception, bool> _shouldExit;
+
+        public OnExceptionSkipOrExitDispatcher(IDispatcher<TMessage> dispatcher, Func<TMessage, Exception, bool> shouldExit)
+        {
+            _dispatcher = dispatcher;
+            _shouldExit = shouldExit;
+        }
+
+        public OnExceptionSkipOrExitDispatcher(IDispatcher<TMessage> dispatcher, bool alwaysExit)
```
Comment (rcknight) [inline src/core/OnExceptionSkipOrExitDispatcher.cs]: Slightly weird if I set alwaysExit to false ... does that mean sometimesExit? Maybe theres a better name like `exitOnException`?

Actually thinking about it, why would you even use this wrapper if you are setting that to false, it does nothing?

File: src/core/ErrorLoggingDIspatcher.cs (hunk 1 above)
Comment (jbradford) [reply src/core/ErrorLoggingDIspatcher.cs]: That was for the retry, for logging an exception it is possible the logging code will want to include properties from the event type.

Comment (rcknight) [reply src/core/ErrorLoggingDIspatcher.cs]: Ah ok

Comment (rcknight) [reply src/core/ErrorLoggingDIspatcher.cs]: btw just noticed the filename for this has a capital I

Comment (jbradford) [reply src/core/ErrorLoggingDIspatcher.cs]: Dammit, I caught that in the class name, missed it on the file

**Review by rcknight — APPROVED** [review]

**Pull request merged**